src/main/java/org/testng/DependencyMap.java

```diff
@@ -61,12 +61,15 @@ public ITestNGMethod getMethodDependingOn(String methodName, ITestNGMethod fromM
     }
     for (ITestNGMethod m : l) {
       // If they are in the same class hierarchy, they must belong to the same instance,
+      Object fromMethodInstance = fromMethod.getInstance();
+      Object mInstance = m.getInstance();
+      if (mInstance != null && fromMethodInstance != null &&
+              fromMethodInstance.getClass().isAssignableFrom(mInstance.getClass())) {
+          return m;
+      } else {
       // otherwise, it's a method depending on a method in a different class so we
       // don't bother checking the instance
-      if (fromMethod.getRealClass().isAssignableFrom(m.getRealClass())) {
```

Comment:
What was the issue?
If I understand the change well, a minimal diff could be:

```java
if (fromMethod.getRealClass().isAssignableFrom(m.getRealClass()) &&
  m.getInstance() != null && fromMethod.getInstance() !== null) {
  return m;
}
```

Reply:
@juherr - The fix was to consider the instance class and not the declaring class which is what `getRealClass()` returns. In cases of inheritance (as in this case), we should have been checking if the method upon which a dependent method was executed for the current instance. TestNG was not doing that. That is what has been changed. See [here](https://stackoverflow.com/a/10028333) for some examples when involving inheritance and `getDeclaringClass`